src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -154,7 +158,40 @@ protected RubyEncoding regexpPrepareEncoding(RubyRegexp regexp, Object matchStri
                         coreExceptions().typeErrorUninitializedRegexp(this));
             }
 
-            if (regexp.encoding == matchStringEncoding) {
+            // This set of checks optimizes for the very common case where the regexp uses the default encoding and
+            // the match string is valid and uses an ASCII-compatible encoding. This extra set of checks is not present
+            // in MRI as of 3.0.2. While they permit an early return of the method, they do not conflict with any of
+            // the branches in the fallback strategy.
+            if (regexp.encoding == Encodings.US_ASCII) {
+                defaultRegexEncodingProfile.enter();
+
+                // The default String encodings are UTF-8 for internal strings and ASCII-8BIT for external strings.
+                // Both encodings are ASCII-compatible and as such can either be CR_7BIT or CR_VALID at this point
+                // depending on the contents. CR_BROKEN strings are handled as a failure case earlier.
+
+                if (codeRange == CodeRange.CR_7BIT) {
+                    asciiOnlyProfile.enter();
+
+                    return Encodings.US_ASCII;
+                } else if (matchStringEncoding == Encodings.UTF_8) {
```

Comment:
We can chat as a group on Slack, but I prefer the longer names. I found the existing regexp code difficult to follow because sometimes code was referring to the pattern string and other times to the match string. You can figure it out from context, but it makes looking at any given method more error-prone.

Reply:
No problem, that's a good reason, keep it as-is, that's why I said `maybe` :)
(Same reasoning for characterIndex vs byteIndex which I find very useful)